File: companion/api/preview.py
"""Live preview streaming to mediamtx."""

import subprocess
import threading
import time
from typing import TYPE_CHECKING, Optional

from config import MEDIAMTX_RTSP_URL
from encoders import _build_rtsp_video_output_args
from logging_setup import get_logger
from process_utils import _spawn_stderr_logger
from process_utils import _terminate_process


if TYPE_CHECKING:
    from config_state import ConfigState
    from managers import MediamtxManager
    from recorder import RecorderState


logger = get_logger()


class PreviewPush:
    """Single dvgrab/iec61883 → ffmpeg → RTSP process for live preview.

    Started lazily when the first MJPEG client connects (or when WebRTC is
    requested) and stopped when recording starts or the API shuts up.
    Only one FireWire source process is ever running at a time.
    """

    # Seconds to wait before retrying after a failure (prevents process storm)
    _RETRY_COOLDOWN = 3.0

    def __init__(
        self,
        config_state: "ConfigState",
        mediamtx_manager: "MediamtxManager",
        recorder_state: "RecorderState",
    ) -> None:
        self._config = config_state
        self._mediamtx = mediamtx_manager
        self._recorder = recorder_state
        self._lock = threading.Lock()
        self._dvgrab: Optional[subprocess.Popen] = None
        self._ffmpeg: Optional[subprocess.Popen] = None
        self._last_failure_ts: float = 0.0

    def ensure_running(self) -> None:
        """Start preview push if not already running and not recording."""
        with self._lock:
            if self._recorder.is_recording:
                return  # Recording owns the bus + mediamtx
            if self._is_alive():
                return

            # Cooldown: if the last attempt failed recently, don't hammer the
            # FireWire bus with new processes — wait for the camera to settle.
            elapsed_since_fail = time.time() - self._last_failure_ts
            if elapsed_since_fail < self._RETRY_COOLDOWN:
                logger.info(
                    "preview-push-cooldown remaining=%.1fs",
                    self._RETRY_COOLDOWN - elapsed_since_fail,
                )
                return

            capture_mode = self._config.get_mode()
            logger.info("preview-push-start capture_mode=%s", capture_mode)

            if not self._mediamtx.is_running():
                self._mediamtx.start()

            rtsp_args = _build_rtsp_video_output_args(MEDIAMTX_RTSP_URL)

            if capture_mode == "dvgrab":
                try:
                    self._dvgrab = subprocess.Popen(
                        ["dvgrab", "--format", "raw", "-"],
                        stdout=subprocess.PIPE,
                        stderr=subprocess.PIPE,
                        bufsize=0,
                        start_new_session=True,
                    )
                    _spawn_stderr_logger(self._dvgrab, "preview-dvgrab")
                except FileNotFoundError:
                    logger.error("preview-push-dvgrab-not-found")
                    return

                try:
                    self._ffmpeg = subprocess.Popen(
                        [
                            "ffmpeg",
                            "-hide_banner",
                            "-loglevel", "error",
                            "-fflags", "nobuffer",
                            "-flags", "low_delay",
                            "-f", "dv",
                            "-i", "pipe:0",
                            *rtsp_args,
                        ],
                        stdin=self._dvgrab.stdout,
                        stdout=subprocess.DEVNULL,
                        stderr=subprocess.PIPE,
                        bufsize=0,
                        start_new_session=True,
                    )
                    _spawn_stderr_logger(self._ffmpeg, "preview-ffmpeg")
                except Exception as e:
                    logger.error("preview-push-ffmpeg-failed error=%s", e)
                    _terminate_process(self._dvgrab)
                    self._dvgrab = None
                    return

                if self._dvgrab.stdout is not None:
                    self._dvgrab.stdout.close()

                # Brief sanity-check: if processes die within 500 ms, camera is absent
                time.sleep(0.5)
                if self._ffmpeg.poll() is not None or self._dvgrab.poll() is not None:
                    logger.error(
                        "preview-push-early-exit dvgrab-rc=%s ffmpeg-rc=%s",
                        self._dvgrab.poll(),
                        self._ffmpeg.poll(),
                    )
                    _terminate_process(self._ffmpeg)
                    _terminate_process(self._dvgrab)
                    self._ffmpeg = None
                    self._dvgrab = None
                    self._last_failure_ts = time.time()
                    return

                logger.info(
                    "preview-push-running dvgrab-pid=%s ffmpeg-pid=%s",
                    self._dvgrab.pid,
                    self._ffmpeg.pid,
                )

            else:  # ffmpeg-only
                try:
                    self._ffmpeg = subprocess.Popen(
                        [
                            "ffmpeg",
                            "-hide_banner",
                            "-loglevel", "error",
                            "-fflags", "nobuffer",
                            "-flags", "low_delay",
                            "-f", "iec61883",
                            "-i", "auto",
                            *rtsp_args,
                        ],
                        stdout=subprocess.DEVNULL,
                        stderr=subprocess.PIPE,
                        bufsize=0,
                        start_new_session=True,
                    )
                    _spawn_stderr_logger(self._ffmpeg, "preview-ffmpeg-direct")
                except Exception as e:
                    logger.error("preview-push-ffmpeg-only-failed error=%s", e)
                    return

                # Brief sanity-check for iec61883 start
                time.sleep(0.5)
                if self._ffmpeg.poll() is not None:
                    logger.error(
                        "preview-push-early-exit ffmpeg-rc=%s (iec61883 unavailable?)",
                        self._ffmpeg.returncode,
                    )
                    self._ffmpeg = None
                    self._last_failure_ts = time.time()
                    return

                logger.info("preview-push-running ffmpeg-pid=%s", self._ffmpeg.pid)

    def stop(self) -> None:
        with self._lock:
            _terminate_process(self._ffmpeg)
            _terminate_process(self._dvgrab)
            self._ffmpeg = None
            self._dvgrab = None
            logger.info("preview-push-stopped")

    def is_alive(self) -> bool:
        with self._lock:
            return self._is_alive()

    def _is_alive(self) -> bool:
        return self._ffmpeg is not None and self._ffmpeg.poll() is None

File: companion/server/internal/stream/preview.go
package stream

import (
	"log/slog"
	"os/exec"
	"sync"
	"time"

	"equip1/companion/server/internal/capture"
	"equip1/companion/server/internal/config"
	"equip1/companion/server/internal/encoders"
	"equip1/companion/server/internal/proc"
)

// previewRetryCooldown is the wait after a failure before retrying, to avoid
// hammering the FireWire bus with new processes.
const previewRetryCooldown = 3 * time.Second

// PreviewPush is the lazy dvgrab|iec61883 → ffmpeg → RTSP live preview. It is
// started lazily (first MJPEG/WebRTC request) and stopped when recording starts.
// Only one FireWire source process is ever running at a time. Ports
// preview.PreviewPush.
type PreviewPush struct {
	captureMode *config.CaptureMode
	mediamtx    *MediamtxManager
	rtspURL     string

	// isRecording is injected to break the recorder↔preview import cycle; it
	// reports whether recording currently owns the bus.
	isRecording func() bool

	mu          sync.Mutex
	dvgrab      *proc.Proc
	ffmpeg      *proc.Proc
	lastFailure time.Time
}

// NewPreviewPush constructs a preview push. SetRecordingCheck must be called to
// wire the recording-state predicate before use (mirrors main.py setting
// preview._recorder after construction).
func NewPreviewPush(captureMode *config.CaptureMode, mediamtx *MediamtxManager, rtspURL string) *PreviewPush {
	return &PreviewPush{
		captureMode: captureMode,
		mediamtx:    mediamtx,
		rtspURL:     rtspURL,
		isRecording: func() bool { return false },
	}
}

// SetRecordingCheck injects the recording-state predicate.
func (p *PreviewPush) SetRecordingCheck(fn func() bool) {
	p.mu.Lock()
	p.isRecording = fn
	p.mu.Unlock()
}

// EnsureRunning starts the preview push if not already running and not recording.
func (p *PreviewPush) EnsureRunning() {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.isRecording() {
		return // recording owns the bus + mediamtx
	}
	if p.aliveLocked() {
		return
	}

	// Cooldown: don't hammer the FireWire bus after a recent failure.
	if elapsed := time.Since(p.lastFailure); elapsed < previewRetryCooldown {
		slog.Info("preview-push-cooldown", "remaining", (previewRetryCooldown - elapsed).Seconds())
		return
	}

	captureMode := p.captureMode.Get()
	slog.Info("preview-push-start", "capture_mode", captureMode)

	if !p.mediamtx.IsRunning() {
		p.mediamtx.Start()
	}

	rtspArgs, err := encoders.BuildRTSPVideoOutputArgs(p.rtspURL)
	if err != nil {
		slog.Error("preview-push-encoder-unavailable", "error", err)
		p.lastFailure = time.Now()
		return
	}

	if captureMode == "dvgrab" {
		p.startDvgrabLocked(rtspArgs)
	} else {
		p.startIec61883Locked(rtspArgs)
	}
}

// startDvgrabLocked spawns dvgrab piped directly into ffmpeg → RTSP. Caller holds p.mu.
func (p *PreviewPush) startDvgrabLocked(rtspArgs []string) {
	dvgrabArgs := capture.DvgrabArgs()
	dvgrabCmd := exec.Command(dvgrabArgs[0], dvgrabArgs[1:]...)

	// dvgrab stdout → ffmpeg stdin, connected directly via one os.Pipe.
	pipeR, pipeW, err := proc.NewStdoutPipe(dvgrabCmd)
	if err != nil {
		slog.Error("preview-push-dvgrab-pipe-failed", "error", err)
		return
	}
	dvErrR, dvErrW, err := proc.NewStderrPipe(dvgrabCmd)
	if err != nil {
		pipeR.Close()
		pipeW.Close()
		return
	}
	dvgrabProc, err := proc.Start(dvgrabCmd)
	if err != nil {
		// dvgrab not found, etc.
		pipeR.Close()
		pipeW.Close()
		dvErrR.Close()
		dvErrW.Close()
		slog.Error("preview-push-dvgrab-not-found", "error", err)
		return
	}
	dvErrW.Close()
	proc.DrainStderr("preview-dvgrab", dvErrR)

	ffArgs := capture.PreviewDvgrabFFmpegArgs(rtspArgs)
	ffmpegCmd := exec.Command(ffArgs[0], ffArgs[1:]...)
	ffmpegCmd.Stdin = pipeR // read directly from dvgrab's stdout pipe
	ffErrR, ffErrW, err := proc.NewStderrPipe(ffmpegCmd)
	if err != nil {
		dvgrabProc.Terminate(3 * time.Second)
		pipeR.Close()
		pipeW.Close()
		return
	}
	ffmpegProc, err := proc.Start(ffmpegCmd)
	if err != nil {
		slog.Error("preview-push-ffmpeg-failed", "error", err)
		dvgrabProc.Terminate(3 * time.Second)
		pipeR.Close()
		pipeW.Close()
		ffErrR.Close()
		ffErrW.Close()
		return
	}
	ffErrW.Close()
	proc.DrainStderr("preview-ffmpeg", ffErrR)

	// Parent closes both pipe ends — they belong to the children now.
	pipeW.Close()
	pipeR.Close()

	p.dvgrab = dvgrabProc
	p.ffmpeg = ffmpegProc

	// Sanity-check: if a process dies within 500ms, the camera is absent.
	time.Sleep(500 * time.Millisecond)
	if p.ffmpeg.Exited() || p.dvgrab.Exited() {
		slog.Error("preview-push-early-exit",
			"dvgrab_rc", p.dvgrab.ExitCode(), "ffmpeg_rc", p.ffmpeg.ExitCode())
		p.ffmpeg.Terminate(3 * time.Second)
		p.dvgrab.Terminate(3 * time.Second)
		p.ffmpeg = nil
		p.dvgrab = nil
		p.lastFailure = time.Now()
		return
	}
	slog.Info("preview-push-running", "dvgrab_pid", p.dvgrab.Pid(), "ffmpeg_pid", p.ffmpeg.Pid())
}

// startIec61883Locked spawns the direct iec61883 → ffmpeg → RTSP push. Caller holds p.mu.
func (p *PreviewPush) startIec61883Locked(rtspArgs []string) {
	ffArgs := capture.PreviewIec61883FFmpegArgs(rtspArgs)
	ffmpegCmd := exec.Command(ffArgs[0], ffArgs[1:]...)
	ffErrR, ffErrW, err := proc.NewStderrPipe(ffmpegCmd)
	if err != nil {
		return
	}
	ffmpegProc, err := proc.Start(ffmpegCmd)
	if err != nil {
		slog.Error("preview-push-ffmpeg-only-failed", "error", err)
		ffErrR.Close()
		ffErrW.Close()
		return
	}
	ffErrW.Close()
	proc.DrainStderr("preview-ffmpeg-direct", ffErrR)
	p.ffmpeg = ffmpegProc

	// Sanity-check for iec61883 start.
	time.Sleep(500 * time.Millisecond)
	if p.ffmpeg.Exited() {
		slog.Error("preview-push-early-exit", "ffmpeg_rc", p.ffmpeg.ExitCode(), "note", "iec61883 unavailable?")
		p.ffmpeg = nil
		p.lastFailure = time.Now()
		return
	}
	slog.Info("preview-push-running", "ffmpeg_pid", p.ffmpeg.Pid())
}

// Stop terminates the preview processes.
func (p *PreviewPush) Stop() {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.ffmpeg != nil {
		p.ffmpeg.Terminate(3 * time.Second)
	}
	if p.dvgrab != nil {
		p.dvgrab.Terminate(3 * time.Second)
	}
	p.ffmpeg = nil
	p.dvgrab = nil
	slog.Info("preview-push-stopped")
}

// IsAlive reports whether the preview ffmpeg is running.
func (p *PreviewPush) IsAlive() bool {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.aliveLocked()
}

func (p *PreviewPush) aliveLocked() bool {
	return p.ffmpeg != nil && !p.ffmpeg.Exited()
}
